spi/README.md

```diff
@@ -16,24 +16,24 @@ public class SampleResourcePlugin extends Plugin implements SystemIndexPlugin, R
 
     @Override
     public Collection<SystemIndexDescriptor> getSystemIndexDescriptors(Settings settings) {
-        final SystemIndexDescriptor systemIndexDescriptor =
-            new SystemIndexDescriptor(RESOURCE_INDEX_NAME, "Sample index with resources");
-        return Collections.singletonList(systemIndexDescriptor);
+      final SystemIndexDescriptor systemIndexDescriptor = new SystemIndexDescriptor(RESOURCE_INDEX_NAME, "Sample index with resources");
+      return Collections.singletonList(systemIndexDescriptor);
     }
 
     @Override
-    public String getResourceType() {
-        return SampleResource.class.getCanonicalName();
+    public Set<ResourceProvider> getResourceProviders() {
+      return Set.of(
+              new ResourceProvider(
+                      SampleResource.class.getCanonicalName(), // class-name of the resource
+                      RESOURCE_INDEX_NAME,                     // the index that stores resource, **must only store the type of resource defined in the line above**
+                      new SampleResourceParser()               // parser to parse the resource
```

Comment:
Are you still able to retain the ResourceProvider but also have a `compileOnly` dependency on the SPI in the extending plugin?

I think the SPI needs to be compileOnly for plugin that optionally extend security.

Reply:
without `implementation` declaration on SPI, there would be RunTimeException (NoClassDefFoundError) and prevent cluster from bootstrapping.

For now these need to be `implementation`, unless we get rid of ShareableResource and ShareableResourceParser interfaces, which are needed for listing all resources API, and without these contracts, resources could be of any type.

Reply:
I have changed it to be compileOnly and converted the listResources API to getResourceIds.